design-documents/graph-ql/coverage/Cart.graphqls

```diff
@@ -0,0 +1,59 @@
+type Query {
+    cart(input: CartQueryInput): CartQueryOutput
+}
+
+input CartQueryInput {
+    cart_id: String!
+}
+
+type CartQueryOutput {
+    cart: Cart
+}
+
+type Cart {
+    id: String!
+
+    line_items_count: Int!
+    items_quantity: Float!
+
+    selected_payment_method: CheckoutPaymentMethod
+    available_payment_methods: [CheckoutPaymentMethod]!
+
+    customer: CheckoutCustomer
+    customer_notes: String
+
+    gift_cards_amount_used: Money
+    applied_gift_cards: [CartGiftCard]
+
+    is_multishipping: Boolean!
+    is_virtual: Boolean!
+}
+
+type CheckoutCustomer {
```

Comment:
It would be great to use polymorphism here.

```
interface CheckoutCustomer {
  # the mandatory fields
}

type GuestCheckoutCustomer implements CheckoutCustomer {
  # …
  middle_name: String
}
type LoggedInCheckoutCustomer implements CheckoutCustomer {
  # …
  middle_name: String!
}
```

React components could then require different fragments depending on the type

Reply:
Good idea in general. I don't understand though, why `middle_name` should be required for the registered customer. The only difference I see right now is that we can eliminate `is_guest` field in this way. 

Please clarify which exact fields should be different between guest and customer implementations.

Reply:
I agree that the `middle_name` is not a good example. It is for mandatory user information upon registration, such as `email` for instance.

The idea of having such a structure, is to open rooms for extension in the shop. For instance in a B2B context, a `LoggedInCheckoutCustomer` may have a `sales_representative: SalesPerson!` attached which would not be available for guests.
Developers would then only extend the `LoggedInCheckoutCustomer` type, making implementation more explicit.